lib/recursive-methods.rb

```diff
@@ -1,49 +1,99 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# source: https://www.bigocheatsheet.com/
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Thank you to Chris-recursion video
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Why can't I use .last?
+def reverse(s, first = 0, last = s.length-1)
+    return s if first > last
+
+    holder = s[first]
+    s[first] = s[last]
+    s[last] = holder
+
+    return reverse(s, first +1, last -1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
+# source: https://medium.com/@txea/reverse-a-string-with-recursion-a3f8505ffb73
 def reverse_inplace(s)
```

Comment:
Since you are making a recursive call with each iteration and you're actually making a new string with the  `s[1..-1]` line.  So this example isn't actually in place.  Although you did the inplace version above.